pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,766 @@
+		captureErr := c.performCapture(ctx, pc, pcState, device)
+		func() {
+			c.mutex.Lock()
+			defer c.mutex.Unlock()
+			c.numRunningCaptures -= 1
+			state := c.captures[pc.Name]
+			if state != nil {
+				state.phase = packetCapturePhaseComplete
+				state.err = captureErr
+			}
+
+		}()
+		c.enqueuePacketCapture(pc)
+	}()
+	return nil
+}
+
+func (c *Controller) performCapture(
+	ctx context.Context,
+	pc *crdv1alpha1.PacketCapture,
+	captureState *packetCaptureState,
+	device string,
+) error {
+	srcIP, dstIP, err := c.parseIPs(ctx, pc)
+	if err != nil {
+		return err
+	}
+	pcapngFile, pcapngWriter, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	updateRateLimiter := rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	packets, err := c.captureInterface.Capture(ctx, device, srcIP, dstIP, pc.Spec.Packet)
+	if err != nil {
+		klog.ErrorS(err, "Failed to start capture")
+		return err
+	}
+	klog.InfoS("Starting capture packets", "name", pc.Name, "device", device)
+	for {
+		select {
+		case packet := <-packets:
+			c.mutex.Lock()
+			captureState.capturedPacketsNum++
+			c.mutex.Unlock()
+			ci := gopacket.CaptureInfo{
+				Timestamp:     time.Now(),
+				CaptureLength: len(packet.Data()),
+				Length:        len(packet.Data()),
+			}
+			err = pcapngWriter.WritePacket(ci, packet.Data())
+			if err != nil {
+				return fmt.Errorf("couldn't write packets: %w", err)
+			}
+			klog.V(5).InfoS("Capture packets", "name", captureState.name, "count",
+				captureState.capturedPacketsNum, "len", ci.Length)
+
+			c.mutex.Lock()
+			reachTarget := captureState.isCaptureSuccessful()
+			c.mutex.Unlock()
+			// if reach the target. flush the file and upload it.
+			if reachTarget {
+				path := env.GetPodName() + ":" + nameToPath(pc.Name)
+				statusPath := path
+				if err = pcapngWriter.Flush(); err != nil {
+					return err
+				}
+				if pc.Spec.FileServer != nil {
+					err = c.uploadPackets(ctx, pc, pcapngFile)
+					klog.V(4).InfoS("Upload captured packets", "name", pc.Name, "path", path)
+					statusPath = fmt.Sprintf("%s/%s.pcapng", pc.Spec.FileServer.URL, pc.Name)
+				}
+				c.mutex.Lock()
+				captureState.filePath = statusPath
+				c.mutex.Unlock()
+				if err != nil {
+					return err
+				}
+				if err := pcapngFile.Close(); err != nil {
+					klog.ErrorS(err, "Close pcapng file error", "name", pc.Name, "path", path)
+				}
+				return nil
+			} else if updateRateLimiter.Allow() {
+				// use rate limiter to reduce the times we need to update status.
+				c.enqueuePacketCapture(pc)
+			}
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (c *Controller) getPodIP(ctx context.Context, podRef *crdv1alpha1.PodReference) (net.IP, error) {
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podRef.Name, podRef.Namespace)
+	var podIP net.IP
+	if len(podInterfaces) > 0 {
+		podIP = podInterfaces[0].GetIPv4Addr()
+	} else {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
```

Comment:
The usage is wrong. If I understand @antoninbas correctly, the context is supposed to be used by the Get call.

Reply:
updated.